feat: Add per-client OAuth scope validation infrastructure

volnei opened the pull request:
## What does this PR do?

This PR implements per-client OAuth scope validation infrastructure to address OAuth scope escalation vulnerabilities. It extends the existing OAuth scope validation (from PR #25336) to support per-client scope restrictions based on app-store manifest configurations.

**Key Changes:**
- Adds `appSlug` field to `OAuthClient` schema to link clients to app-store integrations
- Extends `AppMeta` type with `calOAuthScopes` field for static scope configuration
- Updates app-store CLI to generate static OAuth scope registry
- Implements `parseAndValidateScopesForClient` function for per-client validation
- Updates `generateAuthCode` handler to use per-client scope validation
- Maintains backward compatibility: clients without `appSlug` use basic enum validation

**Architecture:**
- Static configuration via app-store manifests (no database queries for scope lookup)
- Repository pattern for OAuth client data access (no direct Prisma imports outside repositories)
- Defense-in-depth: validation at authorization code generation, with re-validation at token endpoints
- Graceful fallback: clients without `appSlug` or apps without `calOAuthScopes` use basic enum validation

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a documentation change. **N/A** - This is infrastructure for future scope configuration; docs will be added when apps start using `calOAuthScopes`
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works. (29 unit tests for scope validator)

## How should this be tested?

**Environment Setup:**
- No special environment variables required
- Standard Cal.com development setup

**Test Scenarios:**

1. **Backward Compatibility (Critical)**
   - Create an OAuth client without `appSlug` field
   - Verify authorization flow works with basic enum validation
   - Expected: Client can request any valid `AccessScope` enum values

2. **Per-Client Validation (Future)**
   - Add `calOAuthScopes` to an app manifest (e.g., `packages/app-store/googlecalendar/_metadata.ts`)
   - Create an OAuth client with `appSlug` pointing to that app
   - Request scopes outside the allowed list
   - Expected: Authorization fails with "Scopes not allowed for this client" error

3. **Unit Tests**
   - Run: `TZ=UTC yarn test packages/features/auth/lib/__tests__/scopeValidator.test.ts`
   - Expected: All 29 tests pass

4. **Type Check**
   - Run: `yarn type-check:ci --force`
   - Expected: No new type errors (existing errors in workflows/agents/embeds are pre-existing)

**Important Notes:**
- The `oauth.scopes.generated.ts` registry is currently empty because no apps have `calOAuthScopes` defined yet
- Per-client validation will only activate once apps add `calOAuthScopes` to their manifests
- This PR provides the infrastructure; actual scope restrictions will be configured in follow-up PRs

## Human Review Checklist

**Critical Items:**
- [ ] Verify database migration applies cleanly: `packages/prisma/migrations/20251121211030_add_app_slug_to_oauth_client/migration.sql`
- [ ] Test backward compatibility: Existing OAuth clients without `appSlug` should work unchanged
- [ ] Review empty `oauth.scopes.generated.ts` registry - is this expected for infrastructure PR?
- [ ] Verify the dynamic import pattern in `parseAndValidateScopesForClient` is acceptable
- [ ] Check if `parseAndValidateScopesForClient` needs additional unit tests (currently only basic validator is tested)

**Secondary Items:**
- [ ] Review lint fixes in `build.ts` (lines 25, 85) - these were pre-existing warnings, not introduced by this PR
- [ ] Verify repository pattern usage follows Cal.com conventions
- [ ] Check error messages are user-friendly and don't expose internal details

---

**Session:** https://app.devin.ai/sessions/c96cf36a3e6642058414539a22a61fdf  
**Requested by:** Volnei Munhoz (@volnei)

devin-ai-integration[bot] committed:
feat: improve OAuth scope validation and authorization flow

- Add comprehensive scope validation utility
- Validate scopes against AccessScope enum
- Add defense-in-depth validation at token endpoint
- Improve error handling in authorization UI
- Add unit tests for scope validation

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

joeauyeung commented on packages/features/auth/lib/scopeValidator.ts:
```diff
@@ -0,0 +1,55 @@
+      return { success: false, error: "At least one scope is required" };
+    }
+    return { success: true, scopes: [AccessScope.READ_PROFILE] };
+  }
```
This part raises some flags for me. In the places this is being called we're not passing `allowEmpty` and should we really be returning `success` and a scope if no scopes were provided?

devin-ai-integration[bot] committed:
fix: address PR feedback on OAuth scope validation

- Change HTTP status from 401 to 400 for invalid_grant errors per RFC 6749 §5.2
- Pass allowEmpty: false at both validation call sites to make scope requirement explicit
- Ensures token endpoint returns proper OAuth error codes for grant-related failures

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

devin-ai-integration[bot] committed:
feat: add scope validation to refresh token endpoint

- Validate scopes when minting new tokens from refresh tokens
- Use parseAndValidateScopes with allowEmpty: false
- Return 400 with invalid_grant error for invalid scopes per RFC 6749
- Ensures comprehensive scope validation across all OAuth provider endpoints

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

devin-ai-integration[bot] committed:
feat: add per-client OAuth scope validation

- Add appSlug field to OAuthClient schema to link clients to app-store integrations
- Extend AppMeta type with calOAuthScopes field for static scope configuration
- Update app-store CLI to generate static OAuth scope registry
- Create OAuthClientRepository following repository pattern
- Add parseAndValidateScopesForClient function for per-client validation
- Update generateAuthCode handler to use per-client scope validation
- Maintain backward compatibility: clients without appSlug use basic enum validation
- Fix pre-existing lint warnings in build.ts (escape character, require import)
- All unit tests passing (29/29)

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>
